File: 312. Burst Balloons.java
class Solution {//time of O(N^3) and space of O(N^2)
    public int maxCoins(int[] nums) {
        
        int N = nums.length ;
        
        if(N==0) return 0;
        int[][] dp = new int[N][N];
        
        for(int len = 1 ; len<= N ; len++){
            for(int left =0 ; left<= N-len ; left++){
                int right = left + len - 1;
                for(int last=left; last<=right; last++){
                    int leftVal = 1;
                    int rightVal = 1;
                    if(left!=0){
                        leftVal = nums[left-1];
                    }
                    if(right!=N-1){
                        rightVal = nums[right+1];
                    }
                    int beforeWindow = 0;
                    int afterWindow = 0;
                    if(left!=last){
                        beforeWindow = dp[left][last-1];
                    }
                    if(right!= last){
                        afterWindow = dp[last+1][right];
                    }
                    dp[left][right]= Math.max(dp[left][right],beforeWindow + afterWindow+nums[last]*leftVal*rightVal);
                }
            }
        }
        return dp[0][N-1];
    }
}

File: 312-BurstBalloons.cpp
/* 
    Time Complexity                              :  O(N^3)
    Space Complexity                             :  O(N*N)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n,vector<int>(n,0));
        
        for(int len = 1; len <=n; len++) {
            // start index of the subarray
            for(int i=0;i<=n-len;i++) {
                // end index of the subarray
                int j = i + len - 1;
                // iterate between the start and the end index in the array
                for(int k = i;k<=j;k++) {
                    int before = 0;
                    if(k != i) {
                        before = dp[i][k-1];
                    }
                    
                    int after = 0;
                    if(k != j) {
                        after = dp[k+1][j];
                    }
                    
                    int left = 1;
                    if(i != 0) {
                        left = nums[i-1];
                    }
                    
                    int right = 1;
                    if(j != n-1) {
                        right = nums[j+1];
                    }
                    
                    dp[i][j] = max(dp[i][j], before + left * nums[k] * right + after);
                }
            }
        }
        
        return dp[0][n-1];
    }
};